PayID Easy Checkout (#32)
payid-org/rfcs: src/spec/payid-easy-checkout.md

    email: nkramer@ripple.com
    uri: https://www.ripple.com
  -
    ins: I. Simpson
    name: Ian Simpson
    org: Ripple
    street: 315 Montgomery Street
    city: San Francisco
    region: CA
    code: 94104
    country: US
    phone: -----------------
    email: isimpson@ripple.com
    uri: https://www.ripple.com

normative:
    RFC2119:
    RFC2818:
    RFC8446:
    RFC7231:
    RFC7413:
    RFC6265:
    PAYID-URI:
      title: "The 'payid' URI Scheme"
      target: https://tbd.example.com/
      author:
        ins: D. Fuelling
        fullname: David Fuelling
    PAYID-DISCOVERY:
      title: "PayID Discovery"
      author:
        ins: D. Fuelling
        fullname: David Fuelling

informative:
    RFC5988:

--- note_Feedback

This specification is a draft proposal, and is part of the [PayID Protocol](https://payid.org/) initiative. Feedback related to this document should be sent in the form of a Github issue at: https://github.com/payid-org/rfcs/issues.
 
--- abstract
This specification defines the PayID Easy Checkout Protocol, which can be used to allow two parties to transact money in a formalized way for some means of doing business with one another.

> 

Reviewer: Thinking about this abstract -- I wonder if a better abstract would be something that says:

"This spec allows a payment recipient, such as a merchant or a non-profit, to automatically navigate a user to the user's wallet using only a PayID"

The current abstraction is somewhat vague on that point, and seems to imply that this RFC is helping to formalize a payment, but really it's just formalizing "get a user to his wallet, and also send along some payment details".

The primary use-case of this protocol is to define how two parties use a PayID in the context of knowing how to send a user to their wallet to send funds to a recipient (for example, a merchant, charity, etc).   

--- middle

# Terminology

This protocol can be referred to as the `PayId Easy Checkout Protocol`. It uses the following terminology:

> * PayID client: the endpoint that initiates PayID protocol/sending side of the transaction.

Reviewer: In EZ Checkout, doesn't the merchant initiate the protocol (i.e., the receiving side of the transaction)? Actually, it seems like either the sender or receiever could be a client. I suggest we simply define a payID client & server in the abstract, like this:

```suggestion
* PayID client: the endpoint that queries a PayID server using the PayID Protocol.
```

Author: yep, a lot of these terms were just copied and pasted from the PayID spec, and haven't really been changed.  Was going to formalize them later, but I guess it makes more sense to do that before writing

> * PayID server: the endpoint that returns payment account(s) information/receiving side of the transaction (custodial or non-custodial wallets, exchanges, etc).

Reviewer: ```suggestion
* PayID server: the endpoint that returns payment account(s) information.
```

Reviewer: I think we should identify `PayID Server` in the generic sense (there will be one on the sender and one on the receiver, likely).

> * receiver: individual or entity receiving the transaction.

Reviewer: Are there other transactions besides "payments"?

Author: not for this protocol

> * sender: individual or entity originating the transaction/owner of the PayID[PayID-URI][].

Reviewer: This is confusing because both the sender and receiver will own a PayID. If we're wanting to label the PayID that the sender enters on the merchant site, we should call it "Sender's PayID".

```suggestion
* sender: individual or entity originating a payment.
```

> * wallet: the host of the funds of the `sender`; may or may not be custodied.

Reviewer: ```suggestion
* wallet: the holder of funds for the `sender` (may or may not be custodied).
```

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC2119][] and [RFC9174][].

# Introduction

The PayID Easy Checkout Protocol is a minimal protocol designed to provide a set of standard APIs and flows, 
which can be used to transact money between two entities in a way that requires:
* minimal effort for the user initiating the transaction.
* no server-side software specific to PayID or its protocols for servicing the transaction.
* only UI-based solutions.

## Motivation

> The PayID Easy Checkout Protocol aims to correct the current absence of a consistent and broadly adopted pattern for 

Reviewer: This first sentence feels a bit overstated.

paying for goods and services via a digital wallet on an e-commerce website.
Given the ability to assign arbitrary metadata to a PayID as defined in [PayID-Discovery][], there is an opportunity
to standardize the set of interactions between receiver and sender, specifically the process by which a receiver
directs a sender to their digital wallet to complete a payment.

> We believe this protocol will enable a largely improved user experience in e-commerce transactions by reducing the number

Reviewer: ```suggestion
This protocol will enable an improved paying experience by reducing the number
```

of steps a user must take to complete a transaction, creating a consistent and familiar checkout pattern, and lowering

> the barrier to entry use for cryptocurrency novices.

Reviewer: Wondering if we want to mention crypto at all here? I suppose this could be use to make a fiat payment, no?

Author: Yeah, I guess we probably want to keep it general in the RFC.  Realistically though EZ Checkout will probably never be used for fiat payments, since that market seems to be already dominated by paypal and others

> The second priority of PayID Easy Checkout is to limit the engineering effort needed to implement the protocol. 

Reviewer: ```suggestion

```

Author: Why get rid of this? My thinking behind this was that EZ Checkout is EZ because of the low engineering lift needed to conform to protocol, in contrast with full PIC (whatever that is)

Clients wishing to adopt this pattern should only need to implement UI-level changes in order to make the flow function 
as intended, which may aid in expanding overall adoption, further enhancing the protocol's user experience benefits. 

## Design Goals

### Minimal effort for the user initiating the transaction

The PayID Easy Checkout protocol requires a small number of points of data from the user:
* Their PayID

> * Identifying themselves with the wallet they end up redirected to in order to approve transfer of funds to the receiver.

Reviewer: I don't think this is a requirement of the protocol. In the web and mobile cases, the user will potentially already be signed-in to their app. Also, authing into the app is out of scope for EZ Checkout.

Author: agreed

### No server-side software specific to PayID or its protocols for servicing the transaction

> Since the flow of PayID Easy Checkout is predicated on using the PayID Discovery Protocol and then redirecting the 

Reviewer: ```suggestion
Because the flow of PayID Easy Checkout is predicated on using the PayID Discovery Protocol and then redirecting the 
```

> sender away from the site of the receiver, all of the flow can be instrumented in the browser and doesn't require server-side resources. 

Reviewer: Is EZ Checkout browser only?

Author: We should think about this.  A wallet could deeplink to their app, but that would only work on a mobile browser/app.  I guess an app-only wallet could do something like Xumm and show a QR code and listen for the transaction, redirecting back after it has been completed on the phone.  

Seems like it's out of scope for this spec

Apart from a PayID Discovery compliant PayID Server, The PayID Easy Checkout Protocol does not require server-side 
software to be run by either the sender or receiver for a transaction. The PayID server is capable of providing details 
of where to send the user via the PayID Discovery Protocol. Assuming the wallet used by the sender has implemented 
support in their UI for the PayID Easy Checkout Protocol, the sender can be redirected to their wallet within the 
browser to complete their transaction.

# Example Usage

> This section shows the canonical PayID Easy Checkout flow between a hypothetical merchant and customer. The merchant

Reviewer: ```suggestion
This section shows a non-normative example of PayID Easy Checkout between a hypothetical merchant and customer. The merchant
```

> accepts payments at the PayID pay$merchant.com, and the customer has the PayID alice$wallet.com.

Reviewer: ```suggestion
accepts payments using the PayID pay$merchant.com, and the customer controls the PayID alice$wallet.com.
```

## PayID Easy Checkout Initiation

> In this example, the customer might place some items in their online cart on the merchant's website, then choose

Reviewer: ```suggestion
In this example, the customer might place some items in an online shopping cart on the merchant's website, then choose
```

to checkout.  The merchant would then render a form asking for the customer's PayID, as well as a "Checkout with PayID"

> button.  Once the user inputs their PayID alice$wallet.com and clicked the "Checkout with PayID" button, the merchant

Reviewer: ```suggestion
button.  Once the user inputs their PayID alice$wallet.com and clicks the "Checkout with PayID" button, the merchant
```

site begins the PayID Easy Checkout flow.

## PayID Easy Checkout Wallet Discovery

> The merchant site would resolve the customer's PayID to a host as defined in [PAYID-URI][], in this case resolving

Reviewer: ```suggestion
The merchant site would resolve the customer's PayID to a host as defined in [PAYID-DISCOVERY][], in this case resolving
```

Author: Isn't this part of the PayID URI spec? This doesn't really seem like PayID Discovery, because you're just parsing the PayID URI for a host.

> alice$wallet.com to https://wallet.com. The merchant site would then perform PayID Discovery as defined in

Reviewer: ```suggestion
alice$wallet.com to https://wallet.com. The merchant site would then continue PayID Discovery
```

Author: See above.  Having trouble drawing the line between what is defined in PayID URI and PayID Discovery.  Resolving the PayID URI to a host seems to be part of PayID URI, and PayID Discovery is something you can do after you have done that?

Author: Actually, nvm. this makes sense to me now

> [PAYID-DISCOVERY][] to receive a PayID Easy Checkout JRD like this:

Reviewer: ```suggestion
to obtain a PayID JRD like this:
```

    
    GET /.well-known/webfinger?resource=payid%3Aalice%24wallet.com
    Host: wallet.com
    

> If the server has enabled PayID Easy Checkout in their wallet, they would respond with something like this:

Reviewer: ```suggestion
If the sender's PayID server has enabled Easy Checkout, the server would respond with something like this:
```

     
     HTTP/1.1 200 OK
     Access-Control-Allow-Origin: *
     Content-Type: application/jrd+json

     {
       "subject" : "payid:alice$wallet.com",
       "links" :
       [
         {  
           "rel": "https://payid.org/ns/payid-easy-checkout/1.0",
           "template": "https://wallet.com/checkout?amount={amount}&receiverPayId={receiverPayId}&currency={currency}&nextUrl={nextUrl}"
         }
       ]
     }

## Expand Wallet Discovery URL Template
The merchant would parse the PayID Discovery response and iterate over the "links" collection to find the link with 

> the Relation Type of "https://payid.org/ns/payid-easy-checkout/1.0". The site can then do a search and replace on

Reviewer: We should be consistent in usage with the word "site" (probably should be just "merchant"), esp if we want to leave room for an app doing this?

Author: what about defining "merchant" as the actual entity, like Amazon, and defining "merchant front end" to be the merchant's site/app etc, like amazon.com?

Author: calling this the UI, eg "merchant UI" and "wallet UI"

the "template" field value in the link, replacing all occurrences of the predefined query parameter template names with 
the values they want to send to the customer's wallet. One query parameter of note is the "nextUrl" parameter, which
allows the merchant to supply a redirect or callback URL for the sender's wallet to call once the customer has confirmed
the payment.

## Redirect Customer to Their Wallet

> Once the merchant populates the required query parameters in the URL template, they would redirect the customer to 

Reviewer: ```suggestion
Once the merchant populates the required query parameters in the URL template, the merchant would redirect the customer to 
```

> the resulting URL. In this example, the merchant would like to display a "Thank You" page, and replaces `{nextUrl}` with 

Reviewer: The stuff about "thank you" should be moved up into the previous section where you discuss the important of the "nextUrl" param.

`https://wallet.com/checkout?amount=10&receiverPayId=payid%2Apay%24merchant.com&currency=XRP&nextUrl=https://merchant.com/thankyou`.

## Customer Confirms Payment
After the customer clicks the "Pay with PayID" button the merchant's site, and the merchant performs the previous steps,
the customer will be redirected to their wallet at the URL from the previous step.  The wallet front end can

> read the query parameters from the redirect URL and render a confirmation page or modal with all of the required fields

Reviewer: We should define "redirect URL" in the terminology. Maybe something like: 

"**Redirect URL**: The URL that is the result of the PayID Easy Checkout protocol, and which can be used to redirect a client to a wallet corresponding to a particular PayID."

pre-populated.

Once the customer confirms the payment, the wallet would perform a PayID address lookup on the "receiverPayId" query
parameter to get the payment address of the merchant and submit a transaction to the underlying ledger or payment system.
The merchant can then redirect the user back to the URL specified in the "nextUrl" query parameter, which will display
the "Thank You" page of the merchant.

# PayID Easy Checkout Protocol

> The PayID Easy Checkout Protocol can be used to facilitate an end-to-end checkout flow for users transacting

Reviewer: I think we should try to limit the protocol to the machines/software that need to operate the protocol. E.g., maybe instead of "user" we say "client" (which can be a browser, phone, etc). Not sure what to call the "merchant" -- maybe the "recipient"?

Author: Would both the merchant and wallet have a "client"? So something like:
`The PayID Easy Checkout Protocol can be used to facilitate an end-to-end checkout flow between a recipient client and a wallet client?`

The primary benefit of an e-commerce entity using the PayID Easy Checkout Protocol is its ability to automatically
redirect customers to their digital wallet to complete the checkout process. If merchants and payment
processors were instead forced to create one-off integrations with each wallet, implementing this functionality would
quickly become an untenable undertaking. On the other side of the protocol, individual wallets will likely want to
maintain control over resource locations on their individual domains.

In order to meet the technical needs of both wallets and e-commerce receivers, PayID Easy Checkout extends [PAYID-DISCOVERY][]
by defining a new Link Relation type in the PayID metadata returned by a PayID Discovery query.

E-commerce receivers who wish to perform Easy Checkout MUST query the PayID Discovery server to obtain a PayID Easy Checkout
URL. Digital wallets who wish to enable Easy Checkout for their users MUST host a PayID Discovery server and MUST
respond to PayID Discovery queries with an Easy Checkout URL.

E-commerce receivers SHOULD implement fallback measures to complete checkout if a user's wallet does not support PayID Easy Checkout.

> ### Assemble PayID Easy Checkout Discovery URL

Author: good call on these. copied and pasted from the discovery rfc and never changed the numbers
